### ISSUE TITLE
Create Visual vocabularies

### ISSUE THREAD

**Issue by elboyran** [DynaSlum/SateliteImaging#14]
Create BoVW using SURF (default feature extractor) for the 6 training datastores:

- [x] 250 m = 417 px

- [x]  200 m. = 333 px.

- [x] 150 m. = 250 px.

- [x] 100 m. = 167 px.

- [x] 80 m. = 100 px.

- [x] 50 m. = 83 px.

**Comment by elboyran** [comment]
Creating Visual Vocabulary for image training data store # 1 out of 5

Creating Bag-Of-Features.
-------------------------
* Image category 1: BuiltUp
* Image category 2: Mixed
* Image category 3: NonBuiltUp
* Image category 4: Slum
* Selecting feature point locations using the Detector method.
* Extracting SURF features from the selected feature point locations.
** detectSURFFeatures is used to detect key points for feature extraction.

* Extracting features from 1974 images...done. Extracted 1704412 features.

* Keeping 70 percent of the strongest features from each category.

* Balancing the number of features across all image categories to improve clustering.
** Image category 4 has the least number of strongest features: 41049.
** Using the strongest 41049 features from each of the other image categories.

* Using K-Means clustering to create a 250 word visual vocabulary.
* Number of features          : 164196
* Number of clusters (K)      : 250

* Initializing cluster centers...100.00%.
* Clustering...completed 22/100 iterations (~0.75 seconds/iteration)...converged in 22 iterations.

* Finished creating Bag-Of-Features


Encoding images using Bag-Of-Features.
--------------------------------------
* Image category 1: BuiltUp
* Image category 2: Mixed
* Image category 3: NonBuiltUp
* Image category 4: Slum
* Encoding 1974 images...done.
Elapsed time is 358.581138 seconds.
-----------------------------------------------------------------
Creating Visual Vocabulary for image training data store # 2 out of 5

Creating Bag-Of-Features.
-------------------------
* Image category 1: BuiltUp
* Image category 2: Mixed
* Image category 3: NonBuiltUp
* Image category 4: Slum
* Selecting feature point locations using the Detector method.
* Extracting SURF features from the selected feature point locations.
** detectSURFFeatures is used to detect key points for feature extraction.

* Extracting features from 3098 images...done. Extracted 1641367 features.

* Keeping 70 percent of the strongest features from each category.

* Balancing the number of features across all image categories to improve clustering.
** Image category 4 has the least number of strongest features: 45881.
** Using the strongest 45881 features from each of the other image categories.

* Using K-Means clustering to create a 125 word visual vocabulary.
* Number of features          : 183524
* Number of clusters (K)      : 125

* Initializing cluster centers...100.00%.
* Clustering...completed 21/100 iterations (~0.76 seconds/iteration)...converged in 21 iterations.

* Finished creating Bag-Of-Features


Encoding images using Bag-Of-Features.
--------------------------------------
* Image category 1: BuiltUp
* Image category 2: Mixed
* Image category 3: NonBuiltUp
* Image category 4: Slum
* Encoding 3098 images...done.
Elapsed time is 395.494064 seconds.
-----------------------------------------------------------------
Creating Visual Vocabulary for image training data store # 3 out of 5

Creating Bag-Of-Features.
-------------------------
* Image category 1: BuiltUp
* Image category 2: Mixed
* Image category 3: NonBuiltUp
* Image category 4: Slum
* Selecting feature point locations using the Detector method.
* Extracting SURF features from the selected feature point locations.
** detectSURFFeatures is used to detect key points for feature extraction.

* Extracting features from 5585 images...done. Extracted 1559813 features.

* Keeping 70 percent of the strongest features from each category.

* Balancing the number of features across all image categories to improve clustering.
** Image category 2 has the least number of strongest features: 41427.
** Using the strongest 41427 features from each of the other image categories.

* Using K-Means clustering to create a 84 word visual vocabulary.
* Number of features          : 165708
* Number of clusters (K)      : 84

* Initializing cluster centers...100.00%.
* Clustering...completed 25/100 iterations (~0.66 seconds/iteration)...converged in 25 iterations.

* Finished creating Bag-Of-Features


Encoding images using Bag-Of-Features.
--------------------------------------
* Image category 1: BuiltUp
* Image category 2: Mixed
* Image category 3: NonBuiltUp
* Image category 4: Slum
* Encoding 5585 images...done.
Elapsed time is 503.047274 seconds.
-----------------------------------------------------------------
Creating Visual Vocabulary for image training data store # 4 out of 5

Creating Bag-Of-Features.
-------------------------
* Image category 1: BuiltUp
* Image category 2: Mixed
* Image category 3: NonBuiltUp
* Image category 4: Slum
* Selecting feature point locations using the Detector method.
* Extracting SURF features from the selected feature point locations.
** detectSURFFeatures is used to detect key points for feature extraction.

* Extracting features from 12662 images...done. Extracted 1374853 features.

* Keeping 70 percent of the strongest features from each category.

* Balancing the number of features across all image categories to improve clustering.
** Image category 2 has the least number of strongest features: 22674.
** Using the strongest 22674 features from each of the other image categories.

* Using K-Means clustering to create a 63 word visual vocabulary.
* Number of features          : 90696
* Number of clusters (K)      : 63

* Initializing cluster centers...100.00%.
* Clustering...completed 34/100 iterations (~0.36 seconds/iteration)...converged in 34 iterations.

* Finished creating Bag-Of-Features


Encoding images using Bag-Of-Features.
--------------------------------------
* Image category 1: BuiltUp
* Image category 2: Mixed
* Image category 3: NonBuiltUp
* Image category 4: Slum
* Encoding 12662 images...done.
Elapsed time is 774.408225 seconds.
-----------------------------------------------------------------
Creating Visual Vocabulary for image training data store # 5 out of 5

Creating Bag-Of-Features.
-------------------------
* Image category 1: BuiltUp
* Image category 2: Mixed
* Image category 3: NonBuiltUp
* Image category 4: Slum
* Selecting feature point locations using the Detector method.
* Extracting SURF features from the selected feature point locations.
** detectSURFFeatures is used to detect key points for feature extraction.

* Extracting features from 53289 images...done. Extracted 875081 features.

* Keeping 70 percent of the strongest features from each category.

* Balancing the number of features across all image categories to improve clustering.
** Image category 2 has the least number of strongest features: 6630.
** Using the strongest 6630 features from each of the other image categories.

* Using K-Means clustering to create a 50 word visual vocabulary.
* Number of features          : 26520
* Number of clusters (K)      : 50

* Initializing cluster centers...100.00%.
* Clustering...completed 28/100 iterations (~0.12 seconds/iteration)...converged in 28 iterations.

* Finished creating Bag-Of-Features


Encoding images using Bag-Of-Features.
--------------------------------------
* Image category 1: BuiltUp
* Image category 2: Mixed
* Image category 3: NonBuiltUp
* Image category 4: Slum
* Encoding 53289 images...done.
Elapsed time is 1948.952802 seconds.
-----------------------------------------------------------------
DONE.

**Comment by elboyran** [comment]
Try to visualize the histograms of features => new issue.

**Comment by elboyran** [comment]
Redo it for the larger Training sets. Now we don't have explicit validation set!

**Comment by elboyran** [comment]
Creating Visual Vocabulary for image training data store # 1 out of 5

Creating Bag-Of-Features.
-------------------------
* Image category 1: BuiltUp
* Image category 2: Mixed
* Image category 3: NonBuiltUp
* Image category 4: Slum
* Selecting feature point locations using the Detector method.
* Extracting SURF features from the selected feature point locations.
** detectSURFFeatures is used to detect key points for feature extraction.

* Extracting features from 1974 images...done. Extracted 1713545 features.

* Keeping 70 percent of the strongest features from each category.

* Balancing the number of features across all image categories to improve clustering.
** Image category 4 has the least number of strongest features: 39640.
** Using the strongest 39640 features from each of the other image categories.

* Using K-Means clustering to create a 250 word visual vocabulary.
* Number of features          : 158560
* Number of clusters (K)      : 250

* Initializing cluster centers...100.00%.
* Clustering...completed 27/100 iterations (~0.72 seconds/iteration)...converged in 27 iterations.

* Finished creating Bag-Of-Features


Encoding images using Bag-Of-Features.
--------------------------------------
* Image category 1: BuiltUp
* Image category 2: Mixed
* Image category 3: NonBuiltUp
* Image category 4: Slum
* Encoding 1974 images...done.
Elapsed time is 439.609769 seconds.
-----------------------------------------------------------------
Creating Visual Vocabulary for image training data store # 2 out of 5

Creating Bag-Of-Features.
-------------------------
* Image category 1: BuiltUp
* Image category 2: Mixed
* Image category 3: NonBuiltUp
* Image category 4: Slum
* Selecting feature point locations using the Detector method.
* Extracting SURF features from the selected feature point locations.
** detectSURFFeatures is used to detect key points for feature extraction.

* Extracting features from 3098 images...done. Extracted 1641563 features.

* Keeping 70 percent of the strongest features from each category.

* Balancing the number of features across all image categories to improve clustering.
** Image category 4 has the least number of strongest features: 48154.
** Using the strongest 48154 features from each of the other image categories.

* Using K-Means clustering to create a 125 word visual vocabulary.
* Number of features          : 192616
* Number of clusters (K)      : 125

* Initializing cluster centers...100.00%.
* Clustering...completed 27/100 iterations (~0.87 seconds/iteration)...converged in 27 iterations.

* Finished creating Bag-Of-Features


Encoding images using Bag-Of-Features.
--------------------------------------
* Image category 1: BuiltUp
* Image category 2: Mixed
* Image category 3: NonBuiltUp
* Image category 4: Slum
* Encoding 3098 images...done.
Elapsed time is 482.360958 seconds.
-----------------------------------------------------------------
Creating Visual Vocabulary for image training data store # 3 out of 5

Creating Bag-Of-Features.
-------------------------
* Image category 1: BuiltUp
* Image category 2: Mixed
* Image category 3: NonBuiltUp
* Image category 4: Slum
* Selecting feature point locations using the Detector method.
* Extracting SURF features from the selected feature point locations.
** detectSURFFeatures is used to detect key points for feature extraction.

* Extracting features from 5585 images...done. Extracted 1557608 features.

* Keeping 70 percent of the strongest features from each category.

* Balancing the number of features across all image categories to improve clustering.
** Image category 2 has the least number of strongest features: 41647.
** Using the strongest 41647 features from each of the other image categories.

* Using K-Means clustering to create a 84 word visual vocabulary.
* Number of features          : 166588
* Number of clusters (K)      : 84

* Initializing cluster centers...100.00%.
* Clustering...completed 23/100 iterations (~0.74 seconds/iteration)...converged in 23 iterations.

* Finished creating Bag-Of-Features


Encoding images using Bag-Of-Features.
--------------------------------------
* Image category 1: BuiltUp
* Image category 2: Mixed
* Image category 3: NonBuiltUp
* Image category 4: Slum
* Encoding 5585 images...done.
Elapsed time is 608.495775 seconds.
-----------------------------------------------------------------
Creating Visual Vocabulary for image training data store # 4 out of 5

Creating Bag-Of-Features.
-------------------------
* Image category 1: BuiltUp
* Image category 2: Mixed
* Image category 3: NonBuiltUp
* Image category 4: Slum
* Selecting feature point locations using the Detector method.
* Extracting SURF features from the selected feature point locations.
** detectSURFFeatures is used to detect key points for feature extraction.

* Extracting features from 12662 images...done. Extracted 1373076 features.

* Keeping 70 percent of the strongest features from each category.

* Balancing the number of features across all image categories to improve clustering.
** Image category 2 has the least number of strongest features: 22194.
** Using the strongest 22194 features from each of the other image categories.

* Using K-Means clustering to create a 63 word visual vocabulary.
* Number of features          : 88776
* Number of clusters (K)      : 63

* Initializing cluster centers...100.00%.
* Clustering...completed 26/100 iterations (~0.36 seconds/iteration)...converged in 26 iterations.

* Finished creating Bag-Of-Features


Encoding images using Bag-Of-Features.
--------------------------------------
* Image category 1: BuiltUp
* Image category 2: Mixed
* Image category 3: NonBuiltUp
* Image category 4: Slum
* Encoding 12662 images...done.
Elapsed time is 921.729999 seconds.
-----------------------------------------------------------------
Creating Visual Vocabulary for image training data store # 5 out of 5

Creating Bag-Of-Features.
-------------------------
* Image category 1: BuiltUp
* Image category 2: Mixed
* Image category 3: NonBuiltUp
* Image category 4: Slum
* Selecting feature point locations using the Detector method.
* Extracting SURF features from the selected feature point locations.
** detectSURFFeatures is used to detect key points for feature extraction.

* Extracting features from 53289 images...done. Extracted 875891 features.

* Keeping 70 percent of the strongest features from each category.

* Balancing the number of features across all image categories to improve clustering.
** Image category 2 has the least number of strongest features: 6484.
** Using the strongest 6484 features from each of the other image categories.

* Using K-Means clustering to create a 50 word visual vocabulary.
* Number of features          : 25936
* Number of clusters (K)      : 50

* Initializing cluster centers...100.00%.
* Clustering...completed 19/100 iterations (~0.09 seconds/iteration)...converged in 19 iterations.

* Finished creating Bag-Of-Features


Encoding images using Bag-Of-Features.
--------------------------------------
* Image category 1: BuiltUp
* Image category 2: Mixed
* Image category 3: NonBuiltUp
* Image category 4: Slum
* Encoding 53289 images...done.
Elapsed time is 2114.148690 seconds.
-----------------------------------------------------------------
DONE.

**Comment by elboyran** [comment]
The result for 100px80m:

`Creating Visual Vocabulary for image training data store # 1 out of 1

Creating Bag-Of-Features.
-------------------------
* Image category 1: BuiltUp
* Image category 2: NonBuiltUp
* Image category 3: Slum
* Selecting feature point locations using the Detector method.
* Extracting SURF features from the selected feature point locations.
** detectSURFFeatures is used to detect key points for feature extraction.

* Extracting features from 22323 images...done. Extracted 661030 features.

* Keeping 70 percent of the strongest features from each category.

* Balancing the number of features across all image categories to improve clustering.
** Image category 2 has the least number of strongest features: 80365.
** Using the strongest 80365 features from each of the other image categories.

* Using K-Means clustering to create a 10 word visual vocabulary.
* Number of features          : 241095
* Number of clusters (K)      : 10

* Initializing cluster centers...100.00%.
* Clustering...completed 33/100 iterations (~0.59 seconds/iteration)...converged in 33 iterations.

* Finished creating Bag-Of-Features


Encoding images using Bag-Of-Features.
--------------------------------------
* Image category 1: BuiltUp
* Image category 2: NonBuiltUp
* Image category 3: Slum
* Encoding 22323 images...done.
Elapsed time is 705.031849 seconds.
Creating Visual Vocabulary for image testing data store # 1 out of 1

Creating Bag-Of-Features.
-------------------------
* Image category 1: BuiltUp
* Image category 2: NonBuiltUp
* Image category 3: Slum
* Selecting feature point locations using the Detector method.
* Extracting SURF features from the selected feature point locations.
** detectSURFFeatures is used to detect key points for feature extraction.

* Extracting features from 9567 images...done. Extracted 283449 features.

* Keeping 70 percent of the strongest features from each category.

* Balancing the number of features across all image categories to improve clustering.
** Image category 2 has the least number of strongest features: 34243.
** Using the strongest 34243 features from each of the other image categories.

* Using K-Means clustering to create a 10 word visual vocabulary.
* Number of features          : 102729
* Number of clusters (K)      : 10

* Initializing cluster centers...100.00%.
* Clustering...completed 27/100 iterations (~0.27 seconds/iteration)...converged in 27 iterations.

* Finished creating Bag-Of-Features


Encoding images using Bag-Of-Features.
--------------------------------------
* Image category 1: BuiltUp
* Image category 2: NonBuiltUp
* Image category 3: Slum
* Encoding 9567 images...done.
Elapsed time is 218.313246 seconds.
-----------------------------------------------------------------
DONE.`
